front: not-simulated spaceTimeChart: allow projection on invalid train and invalid pathfinding (train schedule)

kmer2016 opened the pull request:
Closes: 
- [x] #13174 
- [x] #13175
- [x] #13176
- [x] #13177

from the meta [#12772](https://github.com/OpenRailAssociation/osrd/issues/12772)

kmer2016 committed:
ui-charts: prepare Manchette for invalid pathfinding display

Add hidePositions props to support displaying trains without valid pathfinding.

Signed-off-by: nncluzu <ngamenichaka@yahoo.fr>

SharglutDev commented on front/src/modules/simulationResult/components/SpaceTimeChartWrapper/useLazyProjectTrains.ts:
```diff
@@ -37,6 +43,7 @@ const useLazyProjectTrains = ({
     Map<TimetableItemId, TrainSpaceTimeData>
   >(new Map());
   const projectionType = useSelector(getProjectionType);
+  const pathSteps = useSelector(getPathSteps);
```
We can't rely on the store's `pathSteps`. They are updated only when editing/creating a train.
But here we want to get the ones from the `trainIdUsedForProjection`.

SharglutDev commented on front/src/modules/simulationResult/components/SpaceTimeChartWrapper/useLazyProjectTrains.ts:
```diff
@@ -45,25 +52,41 @@ const useLazyProjectTrains = ({
-    if (!path) return undefined;
     const opRefs: OperationalPointReference[] = [];
     const opDistances: number[] = [];
+    if (path && operationalPoints && operationalPoints.length > 0) {
```
I don't think we need `path` here. If we have no path, we won't have `operationalPoints` as they come from `PathProperties` that needs a `Pathfinding`.

SharglutDev commented on front/src/modules/simulationResult/components/SpaceTimeChartWrapper/useLazyProjectTrains.ts:
```diff
@@ -45,25 +52,41 @@ const useLazyProjectTrains = ({
+        blocks: path?.blocks ?? [],
+        routes: path?.routes ?? [],
+        track_section_ranges: path?.track_section_ranges ?? [],
+      },
```
What will happen when calling the endpoints needing `path` with these empty values ?
Shouldn't we rather allow `path` to be `undefined` in `TrainProjectionLazyLoaderOptions` to avoid calling the `OccupancyBlocks` endpoints in `TrainOpProjectionLazyLoader` ?

SharglutDev commented on front/public/locales/fr/operational-studies.json:
```diff
@@ -551,6 +551,7 @@
     "timetableOutput": "Table horaire de sortie",
     "toggleManchetteMenu": "Basculer l'affichage du menu de la manchette",
     "toggleWarpedMap": "Basculer l'affichage de la carte déformée",
+    "warpedMapUnavailable": "La carte déformée n'est pas disponible pour les trains qui ne sont pas calculés",
```
I think we use `trains en non calcul` in french ?

SharglutDev commented on front/src/modules/simulationResult/components/SpaceTimeChartWrapper/useLazyProjectTrains.ts:
```diff
@@ -45,25 +52,41 @@ const useLazyProjectTrains = ({
-
     const loader =
       projectionType === 'trackProjection'
         ? new TrainTrackProjectionLazyLoader(options)
```
Shouldn't we deal with the AC `the space-time chart is automatically switched to operational point mode` here ?
We risk to call the "track projection" when there is no pathfinding.

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/SimulationResults/SimulationResults.tsx:
```diff
@@ -194,13 +194,18 @@ const SimulationResults = ({
     <div className="simulation-results" data-testid="simulation-results">
       {/* SIMULATION : SPACE TIME CHART */}
-      {activeBoards.has('std') && projectionData && projectionData.projectedTrains.length > 0 && (
+      {activeBoards.has('std') && shouldDisplaySTD && (
```
Should we even have a condition to display the STD now ?

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/SimulationResults/SimulationResults.tsx:
```diff
@@ -261,7 +274,12 @@ const SimulationResults = ({
-                    projectionLoaderData={projectionData.projectionLoaderData}
+                    projectionLoaderData={
+                      projectionData?.projectionLoaderData || {
+                        totalTrains: 0,
```
Don't know if this is in the ACs but maybe it could be nice to compute this number with all trains that we will see in the STD

kmer2016 committed:
front: allow projection on invalid train

- enable projection button on invalid train schedule/paced train
- Adapt STD to display trains with invalid pathfinding
- hide distances on Manchette for invalid pathfinding trains
- disable track track occupancy

Signed-off-by: nncluzu <ngamenichaka@yahoo.fr>

clarani commented on front/src/applications/operationalStudies/views/Scenario/components/Timetable/PacedTrain/PacedTrainItem.tsx:
You also need to remove this condition in the occurrenceItem actions menu ;)

kmer2016 replied:
Done

clarani commented on front/src/modules/simulationResult/components/SpaceTimeChartWrapper/useLazyProjectTrains.ts:
```diff
@@ -44,28 +40,32 @@ const useLazyProjectTrains = ({
     const opRefs: OperationalPointReference[] = [];
     const opDistances: number[] = [];
+
+    const opsFromProjection = projectionPath.operationalPoints;
```
If you do this change, you can remove the ? on the lines below
```suggestion
    const { operationalPoints = [] } = projectionPath;
```

kmer2016 replied:
Done

clarani commented on front/src/applications/operationalStudies/hooks/usePathProjection.ts:
```diff
@@ -78,17 +89,70 @@ const usePathProjection = (
+
   return useMemo(() => {
     if (pathfinding?.status !== 'success' || !pathProperties) {
-      return undefined;
```
I think we should be able to return undefined, if we have not projected on any train (for instance if the timetable is empty)

kmer2016 replied:
Done, I add an early return 

```typescript
if (!trainIdUsedForProjection || timetableItemsById.size === 0) {
      return undefined;
}
```

clarani commented on front/src/applications/operationalStudies/hooks/usePathProjection.ts:
```diff
@@ -78,17 +89,70 @@ const usePathProjection = (
       projectingOnSimulatedPathException,
     };
-  }, [pathfinding, pathProperties]);
+  }, [pathfinding, pathProperties, operationalPoints]);
```
Some dependencies are missing: `pathUsedForProjection` and `projectingOnSimulatedPathException`

kmer2016 replied:
Done!

clarani commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -146,6 +146,8 @@ const useScenarioData = (scenario: ScenarioResponse, infraId: number) => {
     [projectedTrainsById]
   );
 
+  // TODO Chaka move the folowing useMemo into usePathProjection
```
You forgot to remove the pathUsedForProjection useMemo below

clarani commented on front/src/applications/operationalStudies/types.ts:
```diff
@@ -152,3 +155,26 @@ export type TimetableItemRoundTripGroups = {
+  pathUsedForProjection?: TrainSchedule['path'] | undefined;
+};
+
+export type PathProjectionResult = ProjectionWithPathFinding | ProjectionWithoutPathFinding;
```
Could we change these types like this ?
```suggestion
export type PathProjectionResult = {
  pathUsedForProjection: TrainSchedule['path'];
  operationalPoints: ProjectionOperationalPoint[];
} & (
  | {
      type: 'successful_pathfinding';
      pathfinding: PathfindingResultSuccess;
      geometry: GeoJsonLineString;
      pathUsedForProjection: TrainSchedule['path'];
      projectingOnSimulatedPathException?: boolean;
    }
  | {
      type: 'failed_pathfinding';
    }
);
```

- Even if the pathfinding is not successful, we should have the path of the timetableItem and the Ops corresponding to this path inputs.
- It's easier to differentiate the 2 cases if you introduce a type (it allows to avoid test `'pathfinding' in ...` which are more fragile).

clarani commented on front/src/applications/operationalStudies/helpers/TrainTrackProjectionLazyLoader.ts:
```diff
@@ -21,11 +21,7 @@ import type { ProjectionResult } from './TrainProjectionLazyLoaderAbstract';
 
 export default class TrainTrackProjectionLazyLoader extends TrainProjectionLazyLoaderAbstract {
```
We should not project on Track if we have not a successful pathfinding. It would be nice to add the path in the arguments of this class and to make it compulsory

clarani commented on front/src/applications/operationalStudies/utils.ts:
```diff
@@ -310,6 +312,10 @@ export const isOperationalPointReference = (
   return true;
 };
 
+export const isProjectionWithPathfinding = (
```
This function could be dropped and replace by simpler tests `projectionPath.type === 'successful_pathfinding'`

clarani reviewed (commented):
Nice work ! I have started to review but I haven't got the time to read all the files yet

SharglutDev commented on front/src/applications/operationalStudies/hooks/usePathProjection.ts:
```diff
@@ -78,17 +89,70 @@ const usePathProjection = (
+      const normalizedOps: ProjectionOperationalPoint[] = rawOps.map((op, index) => ({
+        id: op.id,
+        extensions: op.extensions,
+        part: op.parts[0] || { track: '', position: 0, extensions: undefined },
```
nit

```suggestion
        part: op.parts.at(0) || { track: '', position: 0, extensions: undefined },
```
But is it even possible to have an empty array ? Wouldn't this field have been optional if it was the case ? Or did you encounter that issue ?

SharglutDev commented on front/src/modules/simulationResult/components/SpaceTimeChartWrapper/useLazyProjectTrains.ts:
```diff
@@ -44,28 +40,32 @@ const useLazyProjectTrains = ({
     const opRefs: OperationalPointReference[] = [];
     const opDistances: number[] = [];
+
+    const opsFromProjection = projectionPath.operationalPoints;
```
We could add a condition to fill these 2 arrays only if we are on op projection

SharglutDev commented on front/src/applications/operationalStudies/hooks/usePathProjection.ts:
```diff
@@ -78,17 +89,70 @@ const usePathProjection = (
+        id: op.id,
+        extensions: op.extensions,
+        part: op.parts[0] || { track: '', position: 0, extensions: undefined },
+        position: index * FALLBACK_DISTANCE_MM,
```
What about if we have a train with origin added by map click and an op ref destination ? Don't we want to see the destination in the manchette ? (@Tguisnet)

With this logic, position will be 0 when we would want 100km

Tguisnet replied:
Yes, we want to exclude all clicked point, and keep all op ref, so in this case, we want to see the destination in the manchette, even if it's the only point :) !

SharglutDev replied:
So it would be displayed at the top of the manchette, it's fine like that ?

Tguisnet replied:
Yes

SharglutDev commented on front/src/modules/simulationResult/components/SpaceTimeChartWrapper/SpaceTimeChartWrapper.tsx:
```diff
@@ -138,6 +139,7 @@ const SpaceTimeChartWrapper = ({
   timetableItemsWithDetails,
   waypointsPanelIsOpen,
   setWaypointsPanelIsOpen,
+  pathfindingIsSuccess = true,
```
nit: we usually have default to `false`, we could use a prop like `pathfindingHasFailed`

SharglutDev reviewed (commented):
Great improvement !

Uriel-Sautron commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -341,18 +322,32 @@ const useScenarioData = (scenario: ScenarioResponse, infraId: number) => {
-              },
-            }
-          : undefined,
+      projectionData: projectionPath?.pathUsedForProjection
```
Perhaps putting this logic in a function would make it more readable.

kmer2016 replied:
Good point! I initially tried extracting this logic into a separate function, but after testing both approaches, I think the inline version is actually clearer in this case. The conditional logic is straightforward and only used once, so the abstraction doesn't add much value.

I think sometimes the simplest solution is the best one. What do you think?

clarani replied:
If the props of projectionPath are correctly named, you can simply do:
```
      projectionData: projectionPath
        ? {
            ...projectionPath,
            projectedTrains,
            projectionLoaderData: {
              allTrainsProjected,
              totalTrains: timetableItems?.length ?? 0,
            },
          }
        : undefined,
```

To do so, you need to rename `pathUsedForProjection` -> `path` and remove `pathfindingHasFailed` to keep the type of the projection

Uriel-Sautron commented on front/src/applications/stdcm/hooks/useProjectedTrainsForStdcm.ts:
```diff
@@ -91,7 +91,9 @@ const useProjectedTrainsForStdcm = (stdcmResponse?: StdcmSuccessResponse) => {
     infraId,
     electricalProfileSetId,
-    path: stdcmResponse?.pathfinding_result,
+    projectionPath: {
```
Here projectionPath does not match the type PathProjectionResult

kmer2016 commented:
When creating paced train via NGE with points not in the infrastructure, the backend returns empty arrays. A bug has been created: https://github.com/OpenRailAssociation/osrd/issues/13770

clarani commented on front/src/applications/operationalStudies/hooks/usePathProjection.ts:
```diff
@@ -78,17 +90,125 @@ const usePathProjection = (
+
   return useMemo(() => {
-    if (pathfinding?.status !== 'success' || !pathProperties) {
+    if (!trainIdUsedForProjection || timetableItemsById.size === 0) {
```
I think it's enough to check if `pathUsedForProjection` is `undefined` or not (since it will be undefined only if `!trainIdUsedForProjection || timetableItemsById.size === 0`.
If you do so, you can also change the type `PathProjectionResult` since `pathUsedForProjection` will always be defined

clarani commented on front/src/applications/operationalStudies/types.ts:
```diff
@@ -154,3 +155,19 @@ export type StudyCardDetails = SearchResultItemStudy | StudyWithScenarios;
+      type: 'successful_pathfinding';
+      pathfinding: PathfindingResultSuccess;
+      geometry: PathProperties['geometry'];
+      pathUsedForProjection: PathItem[] | undefined;
```
You can remove `pathUsedForProjection` since you've already declare it above
```suggestion
```

clarani commented on front/src/applications/operationalStudies/views/Scenario/components/SimulationResults/SimulationResults.tsx:
```diff
@@ -104,6 +110,12 @@ const SimulationResults = ({
     }
   }, [projectionData, timetableItemsWithDetails]);
 
+  useEffect(() => {
```
Hum I'm wondering if it's the best place for this `useEffect`. Maybe it can be moved to `useLazyProject`. Or `usePathProjection` ?

SharglutDev replied:
`usePathProjection` looks like a good option.

clarani commented on front/src/modules/simulationResult/components/SpaceTimeChartWrapper/useLazyProjectTrains.ts:
```diff
@@ -45,29 +45,35 @@ const useLazyProjectTrains = ({
-        ? new TrainTrackProjectionLazyLoader(options)
-        : new TrainOpProjectionLazyLoader(opRefs, opDistances, options);
+      operationalPoints.forEach(({ id, position }, index) => {
+        opRefs.push({ operational_point: id });
```
:warning: Here, you only handle the case where the op are found in the infra : if they're not found, they will be filtered in useProjectPath (and they won't have any operational_point id here)

SharglutDev replied:
And regarding @flomonster [comment](https://github.com/OpenRailAssociation/osrd/pull/13492#issuecomment-3435667135), looks like we should always pass the path input locations (the op ref ones) as an input for project path op anyway.

flomonster commented:
Please refer to this comment : https://github.com/OpenRailAssociation/osrd/issues/13770#issuecomment-3435663569

The frontend is currently doing invalid and useless operations.

SharglutDev commented on front/src/applications/operationalStudies/hooks/usePathProjection.ts:
```diff
@@ -78,17 +90,125 @@ const usePathProjection = (
+    // FAILED PATHFINDING HANDLING
+    // ===========================
+    // When pathfinding fails or path properties are unavailable, we still need to display
+    // operational points in the manchette and allow projection in the GET (Graph Espace-Temps).
```
```suggestion
    // operational points in the manchette and allow projection in the STD (space time diagram).
```

SharglutDev commented on front/src/applications/operationalStudies/hooks/usePathProjection.ts:
```diff
@@ -78,17 +90,125 @@ const usePathProjection = (
+            ('trigram' in opRef && opRef.trigram) ||
+            ('uic' in opRef && opRef.uic?.toString()) ||
+            ('operational_point' in opRef && opRef.operational_point) ||
+            `Point ${normalizedOps.length + 1}`;
```
For requested point, we use the `requestedPoint` key in `operational-studies.json` (different key for origin and destination)

SharglutDev commented on front/src/applications/operationalStudies/hooks/usePathProjection.ts:
```diff
@@ -78,17 +90,125 @@ const usePathProjection = (
+            extensions: {
+              identifier: {
+                name: virtualName,
+                uic: ('uic' in opRef && opRef.uic) || 0, // Use actual UIC if available, fallback to 0
```
Why populate only the `uic` and not keep all its existing (even if invalid) inputs ? Like the `trigram`, `secondary_code`, `operational_point`, `track_reference` if available

SharglutDev commented on front/src/modules/simulationResult/components/SpaceTimeChartWrapper/WaypointsPanel.tsx:
We hide the position of ops in the manchette when no pathfinding. Do we want to do the same in the `WaypointsPanel` ?

kmer2016 replied:
Good point, I think yes, we shoulds hide the offset too in WaypointsPanel. @Tguisnet ?

Tguisnet replied:
For now, yes

SharglutDev commented on front/src/applications/operationalStudies/hooks/usePathProjection.ts:
```diff
@@ -171,10 +202,17 @@ const usePathProjection = (
               },
             },
-            part: { track: '', position: 0, extensions: undefined },
+            part: { track: '', position: 0 },
```
Wouldn't the `opRef.track` be useful when we will want to use the TOD when no pathfinding ?

SharglutDev commented on front/src/applications/operationalStudies/types.ts:
```diff
@@ -154,3 +156,23 @@ export type StudyCardDetails = SearchResultItemStudy | StudyWithScenarios;
 export type CategoryColors = { normal: string; hovered: string; background: string };
+
+export type PathProjectionResult = {
+  path: PathItem[] | undefined;
```
Looks like `path` is always defined now
```suggestion
  path: PathItem[];
```

SharglutDev commented on front/src/applications/operationalStudies/types.ts:
```diff
@@ -154,3 +156,23 @@ export type StudyCardDetails = SearchResultItemStudy | StudyWithScenarios;
+      type: 'failed_pathfinding';
+      pathfinding: undefined;
+      geometry: undefined;
+      projectingOnSimulatedPathException: undefined;
```
Can't we remove these ?
```suggestion
```

SharglutDev commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -341,18 +328,18 @@ const useScenarioData = (scenario: ScenarioResponse, infraId: number) => {
-              },
-            }
-          : undefined,
+      projectionData: projectionPath?.path
```
If you made the type change from the above comment, you should be able to do that
```suggestion
      projectionData: projectionPath
```

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/ScenarioContent.tsx:
```diff
@@ -186,7 +186,9 @@ const ScenarioContent = ({ activeBoards }: ScenarioContentProps) => {
-                projectionData?.projectingOnSimulatedPathException
+                projectionData?.type === 'successful_pathfinding'
+                  ? projectionData.projectingOnSimulatedPathException
+                  : undefined
```
```suggestion
              projectingOnSimulatedPathException={
                projectionData?.projectingOnSimulatedPathException
              }
```

SharglutDev commented on front/src/modules/simulationResult/components/SpaceTimeChartWrapper/WaypointsPanel.tsx:
```diff
@@ -16,13 +16,15 @@ type WaypointsPanelProps = {
   setWaypointsPanelIsOpen,
   waypoints,
   waypointsPanelData: { filteredWaypoints, setFilteredWaypoints, projectionPath, timetableId },
+  showOffsets = true,
```
nit: Same as before, would prefer like that

```suggestion
  hideOffsets = false,
```

clarani commented on front/src/applications/operationalStudies/views/Scenario/components/SimulationResults/SimulationResults.tsx:
```diff
@@ -194,13 +195,17 @@ const SimulationResults = ({
   };
 
-  if (!simulationResults && !projectionData) {
+  if (!simulationResults || !projectionData) {
```
Hum do you really want to do this ? If the `projectionData` is loading, you won't have all the simulation results

clarani commented on front/src/modules/simulationResult/components/SpaceTimeChartWrapper/useLazyProjectTrains.ts:
```diff
@@ -20,15 +20,19 @@ import { useAppDispatch } from 'store';
   electricalProfileSetId,
   path,
-  operationalPoints,
+  operationalPoints = [],
```
You don't need `operationalPoints` now that you have `operationalPointReferences`
```suggestion
```

clarani commented on front/src/modules/simulationResult/components/SpaceTimeChartWrapper/useLazyProjectTrains.ts:
```diff
@@ -76,7 +80,7 @@ const useLazyProjectTrains = ({
       loaderRef.current = null;
     };
-  }, [infraId, electricalProfileSetId, path, projectionType, operationalPoints]);
+  }, [infraId, electricalProfileSetId, projectionType, path, operationalPoints]);
```
I think you should add `operationalPointReferences` and `operationalPointDistances` in your dependencies

clarani commented on front/src/applications/operationalStudies/hooks/usePathProjection.ts:
The logic of the hook `usePathProjection` looks good to me :partying_face: I just need to test !

clarani commented on front/src/applications/operationalStudies/hooks/usePathProjection.ts:
```diff
@@ -78,17 +99,169 @@ const usePathProjection = (
+    return exception?.path_and_schedule?.path ?? pacedTrain!.path;
+  }, [trainIdUsedForProjection, timetableItemsById]);
+
+  const requestedOperationalPoints: OperationalPointReference[] = [];
```
I think it's more explicit like this
```suggestion
  const opRefs: OperationalPointReference[] = [];
```

clarani commented on front/src/applications/operationalStudies/hooks/usePathProjection.ts:
```diff
@@ -78,17 +99,169 @@ const usePathProjection = (
+              ? t('main.requestedOrigin')
+              : index === requestedOperationalPoints.length - 1
+                ? t('main.requestedDestination')
+                : t('main.requestedPoint', { count: index + 1 }));
```
Maybe we can hide the details of these computations in a function ?

clarani commented on front/src/applications/operationalStudies/hooks/usePathProjection.ts:
```diff
@@ -78,17 +99,169 @@ const usePathProjection = (
+                ch: ('secondary_code' in opRef && opRef.secondary_code) || '',
+                ch_long_label: '',
+                ch_short_label: '',
+                ci: ('uic' in opRef && opRef.uic) || 0,
```
`uic` is not the same thing than `ci`, you need to truncate the 2 first digits I think

SharglutDev reviewed (commented):
Current state lgtm (waiting to see if more changes are needed).

Just one thought with this example : 

<img width="934" height="570" alt="Image" src="https://github.com/user-attachments/assets/a6838099-b2ab-44c2-aa79-1e98ecada2e7" />

Maybe it's not necessary to display the invalid operational_point id. We could just display "Identifiant de point" / "Operational point identifier" like we are going to do in the itinerary modal (@Tguisnet)

clarani commented on front/src/applications/operationalStudies/hooks/usePathProjection.ts:
```diff
@@ -78,17 +134,158 @@ const usePathProjection = (
+    return exception?.path_and_schedule?.path ?? pacedTrain!.path;
+  }, [trainIdUsedForProjection, timetableItemsById]);
+
+  const opRefs: OperationalPointReference[] = [];
```
Do you think we can useMemo opRefs ? Otherwise it is recomputed at each re-render

clarani commented on front/src/applications/operationalStudies/hooks/usePathProjection.ts:
```diff
@@ -78,17 +134,158 @@ const usePathProjection = (
+    // SUCCESSFUL PATHFINDING
+    // ===========================
+    // Use backend-provided operational points with accurate positions and geometry
+
```
nit: I would handle the successful pathfindings before the invalid ones since this case is easier and simpler

clarani commented on front/src/applications/operationalStudies/hooks/usePathProjection.ts:
```diff
@@ -78,17 +134,158 @@ const usePathProjection = (
+            },
+            part: { track: '', position: 0 },
+            position,
+            weight,
```
To make the code a little easier to read, maybe we can extract this code in a small function `createVirtualOp` :)

clarani commented on front/src/applications/operationalStudies/hooks/usePathProjection.ts:
```diff
@@ -1,21 +1,71 @@
+    return t('main.requestedDestination');
+  }
+
+  return t('main.requestedPoint', { count: index + 1 });
```
nit: you can make this function more compact & readable like this
```suggestion
  if ('trigram' in opRef && opRef.trigram) return opRef.trigram;
  if ('uic' in opRef && opRef.uic) return opRef.uic.toString();
  if ('operational_point' in opRef && opRef.operational_point) return opRef.operational_point;

  if (index === 0) return t('main.requestedOrigin');
  if (index === totalCount - 1) return t('main.requestedDestination');
  return t('main.requestedPoint', { count: index + 1 });
```

clarani commented on front/src/modules/simulationResult/components/SpaceTimeChartWrapper/SpaceTimeChartWrapper.tsx:
```diff
@@ -189,7 +191,7 @@ const SpaceTimeChartWrapper = ({
     useMemo(
       () => cutSpaceTimeChart(projectedTrains, conflicts, operationalPoints, waypointsPanelData),
-      [waypointsPanelData?.filteredWaypoints, projectedTrains, conflicts]
+      [waypointsPanelData?.filteredWaypoints, projectedTrains, conflicts, operationalPoints]
```
Thanks for the fix :pray:

clarani commented on front/src/applications/operationalStudies/hooks/usePathProjection.ts:
```diff
@@ -78,17 +154,138 @@ const usePathProjection = (
-  }, [pathfinding, pathProperties]);
+  }, [
+    trainIdUsedForProjection,
+    timetableItemsById,
```
I think you can remove these 2 props
```suggestion
```

clarani commented on front/ui/ui-charts/src/manchette/hooks/useManchetteWithSpaceTimeChart.tsx:
You don't need to add hidePositions in this hook, you can directly give this props to the Manchette in the html of the SpaceTimeChartWrapper

clarani commented on front/src/modules/simulationResult/types.ts:
```diff
@@ -86,13 +86,14 @@ export type SpeedDistanceDiagramData = {
     totalTrains: number;
   };
   operationalPoints: PathProperties['operational_points'];
+  pathfindingHasFailed: boolean;
```
nit: if the pathfinding has succeeded, then you know that the pathfinding and the geometry are defined
```
export type ProjectionData = { ... } & (
  | {
      pathfindingHasFailed: false;
      pathfinding: PathfindingResultSuccess;
      geometry: PathProperties['geometry'];
    }
  | {
      pathfindingHasFailed: true;
      pathfinding: undefined;
      geometry: undefined;
    }
)
```

clarani commented on front/src/modules/simulationResult/components/SpaceTimeChartWrapper/SpaceTimeChartWrapper.tsx:
```diff
@@ -311,6 +314,7 @@ const SpaceTimeChartWrapper = ({
     defaultTimeOrigin: 0,
     defaultSpaceOrigin:
       (waypointsPanelData?.filteredWaypoints ?? operationalPoints).at(0)?.position || 0,
+    hidePositions: pathfindingHasFailed,
```
No need to give the `hidePositions` boolean to this hook
```suggestion
```

Tguisnet commented:
> Current state lgtm (waiting to see if more changes are needed).
> 
> Just one thought with this example :
> 
> <img alt="Image" width="934" height="570" src="https://private-user-images.githubusercontent.com/113104137/507007242-a6838099-b2ab-44c2-aa79-1e98ecada2e7.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjIxNjI5MzIsIm5iZiI6MTc2MjE2MjYzMiwicGF0aCI6Ii8xMTMxMDQxMzcvNTA3MDA3MjQyLWE2ODM4MDk5LWIyYWItNDRjMi1hYTc5LTFlOThlY2FkYTJlNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMTAzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTEwM1QwOTM3MTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iN2E3ZGM3MzVjMzJhMGIwZjgyZTQ4NDE4Nzc5MmJlMzBlOWQwNWViOTE3M2NkOGJkZDA4MDZkOTA1OTgzZGQyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.sbHYkDNs__ju6R9PoAzQCPcjeCVZ6gKKUOcf1qSgaJs">
> Maybe it's not necessary to display the invalid operational_point id. We could just display "Identifiant de point" / "Operational point identifier" like we are going to do in the itinerary modal (@Tguisnet)

Until you find a better solution, you can use Paul's solution ("Identifiant de point" / "Operational point identifier").

clarani reviewed (commented):
LGTM & tested :white_check_mark: 

Can you clean your git history and fix the e2e tests ?

SharglutDev reviewed (approved):
Thanks for fix in the manchette !

clarani reviewed (commented):
LGTM :white_check_mark:

clarani commented on front/src/applications/operationalStudies/hooks/usePathProjection.ts:
Nit: you can merge the last commit in the 2nd one since they are linked (the CI would fail on the 2nd commit without the 3rd one)

SharglutDev commented:
You should rebase your branch, `deleted` has been removed from `PathItem` type.

Not sure why you added the cast in the `front: fix ci` commit ? Typescript doesn't say anything when I remove it.
